controllers/neutronapi_controller.go

```diff
@@ -89,327 +98,467 @@ func (r *NeutronAPIReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 }
 
-// SetupWithManager -
-func (r *NeutronAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&neutronv1beta1.NeutronAPI{}).
-		Owns(&appsv1.Deployment{}).
-		Complete(r)
+func (r *NeutronAPIReconciler) reconcileUpdate(ctx context.Context, instance *neutronv1beta1.NeutronAPI, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service update")
+
+	// TODO: should have minor update tasks if required
+	// - delete dbsync hash from status to rerun it?
+
+	r.Log.Info("Reconciled Service update successfully")
+	return ctrl.Result{}, nil
 }
 
-func (r *NeutronAPIReconciler) setDbSyncHash(api *neutronv1beta1.NeutronAPI, hashStr string) error {
+func (r *NeutronAPIReconciler) reconcileUpgrade(ctx context.Context, instance *neutronv1beta1.NeutronAPI, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service upgrade")
 
-	if hashStr != api.Status.DbSyncHash {
-		api.Status.DbSyncHash = hashStr
-		if err := r.Client.Status().Update(context.TODO(), api); err != nil {
-			return err
-		}
-	}
-	return nil
+	// TODO: should have major version upgrade tasks
+	// -delete dbsync hash from status to rerun it?
+
+	r.Log.Info("Reconciled Service upgrade successfully")
+	return ctrl.Result{}, nil
 }
 
-func (r *NeutronAPIReconciler) setAPIEndpoint(instance *neutronv1beta1.NeutronAPI, endpoint string) error {
+func (r *NeutronAPIReconciler) reconcileNormal(ctx context.Context, instance *neutronv1beta1.NeutronAPI, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service")
 
-	if endpoint != instance.Status.APIEndpoint {
-		instance.Status.APIEndpoint = endpoint
-		if err := r.Client.Status().Update(context.TODO(), instance); err != nil {
-			return err
+	// If the service object doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(instance, helper.GetFinalizer())
+	// Register the finalizer immediately to avoid orphaning resources on delete
+	//if err := patchHelper.Patch(ctx, openStackCluster); err != nil {
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+	// ConfigMap
+	configMapVars := make(map[string]common.EnvSetter)
+
+	//
+	// check for required OpenStack secret holding passwords for service/admin user and add hash to the vars map
+	//
+	ospSecret, hash, err := common.GetSecret(ctx, helper, instance.Spec.Secret, instance.Namespace)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			return ctrl.Result{RequeueAfter: time.Second * 10}, fmt.Errorf("OpenStack secret %s not found", instance.Spec.Secret)
 		}
+		return ctrl.Result{}, err
 	}
-	return nil
+	configMapVars[ospSecret.Name] = common.EnvValue(hash)
 
-}
+	// check for required ovn-connection configMap
+	ovnConnection, hash, err := common.GetConfigMapAndHashWithName(ctx, helper, instance.Spec.OVNConnectionConfigMap, instance.Namespace)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: time.Second * 10}, err
+	}
```

Comment:
In the CRD definition, `OVNConnectionConfigMap` is not flagged as required.  Should this only check for the config map if `OVNConnectionConfigMap` is not the empty string?  Or should the CRD be changed so that this indeed required?

Reply:
For functionality wise it's needed, so i think would make it as required would be better.
But considering the Data in this config map comes from ovn Resources, so when ovn-operator will be ported, these can be set in OVN resource Status and those can be used instead of using config map.

Currently i used https://github.com/openstack-k8s-operators/osp-director-dev-tools/tree/master/ansible/roles/cnosp/files/ovn for setting up ovn resources and create OVNConnectionConfigMap based on the setup.

Reply:
Understood.  I see you have a PR for Neutron for `install_yamls` as well.  It sounds like the `OVN` automation from `osp-director-dev-tools` might need to be added to `install_yamls` as well.